tutorpod_autoscaling/plugin.py

```diff
@@ -29,76 +30,114 @@
-        "LMS_WORKER_CPU_LIMIT": 1,
-        "LMS_WORKER_CPU_REQUEST": 0.175,
-        "LMS_WORKER_MAX_REPLICAS": LMS_MAX_REPLICAS * 1.5,
-        "LMS_WORKER_MEMORY_LIMIT": f"{LMS_WORKER_MEMORY_REQUEST_MB * 4}Mi",
-        "LMS_WORKER_MEMORY_REQUEST": f"{LMS_WORKER_MEMORY_REQUEST_MB}Mi",
-        "LMS_WORKER_MIN_REPLICAS": 1,
-        # Kubernetes VPA autoscaling settings
-        "LMS_VPA": False,
-        "LMS_WORKER_VPA": False,
-        "CMS_VPA": False,
-        "CMS_WORKER_VPA": False,
     },
     "unique": {},
     "overrides": {},
 }
 
-# Add configuration entries
-hooks.Filters.CONFIG_DEFAULTS.add_items(
-    [(f"POD_AUTOSCALING_{key}", value) for key, value in config["defaults"].items()]
-)
-hooks.Filters.CONFIG_UNIQUE.add_items(
-    [(f"POD_AUTOSCALING_{key}", value) for key, value in config["unique"].items()]
-)
-hooks.Filters.CONFIG_OVERRIDES.add_items(list(config["overrides"].items()))
+CORE_AUTOSCALING_CONFIG: dict[str, AUTOSCALING_ATTRS_TYPE] = {
+    "lms": {
+        "enable_hpa": True,
+        "memory_request": f"{LMS_MEMORY_REQUEST_MB}Mi",
+        "cpu_request": 0.25,
+        "memory_limit": f"{LMS_MEMORY_REQUEST_MB * 4}Mi",
+        "cpu_limit": 1,
+        "min_replicas": 1,
+        "max_replicas": LMS_MAX_REPLICAS,
+        "avg_cpu": 300,
+        "avg_memory": "",
+        "enable_vpa": False,
+    },
+    "lms-worker": {
+        "enable_hpa": True,
+        "memory_request": f"{LMS_WORKER_MEMORY_REQUEST_MB}Mi",
+        "cpu_request": 0.175,
+        "memory_limit": f"{LMS_WORKER_MEMORY_REQUEST_MB * 4}Mi",
+        "cpu_limit": 1,
+        "min_replicas": 1,
+        "max_replicas": LMS_MAX_REPLICAS * 1.5,
+        "avg_cpu": 400,
+        "avg_memory": "",
+        "enable_vpa": False,
+    },
+    "cms": {
+        "enable_hpa": True,
+        "memory_request": f"{CMS_MEMORY_REQUEST_MB}Mi",
+        "cpu_request": 0.25,
+        "memory_limit": f"{CMS_MEMORY_REQUEST_MB * 4}Mi",
+        "cpu_limit": 1,
+        "min_replicas": 1,
+        "max_replicas": CMS_MAX_REPLICAS,
+        "avg_cpu": 300,
+        "avg_memory": "",
+        "enable_vpa": False,
+    },
+    "cms-worker": {
+        "enable_hpa": True,
+        "memory_request": f"{CMS_WORKER_MEMORY_REQUEST_MB}Mi",
+        "cpu_request": 0.175,
+        "memory_limit": f"{CMS_WORKER_MEMORY_REQUEST_MB * 4}Mi",
+        "cpu_limit": 1,
+        "min_replicas": 1,
+        "max_replicas": CMS_MAX_REPLICAS * 1.5,
+        "avg_cpu": 400,
+        "avg_memory": "",
+        "enable_vpa": False,
+    },
+}
+
+# The core autoscaling configs are added with a high priority, such that other users can override or
+# remove them.
+@AUTOSCALING_CONFIG.add(priority=tutor_hooks.priorities.HIGH)
+def _add_core_autoscaling_config(scaling_config: dict[str, AUTOSCALING_ATTRS_TYPE]) -> dict[str, AUTOSCALING_ATTRS_TYPE]:
+    scaling_config.update(CORE_AUTOSCALING_CONFIG)
+    return scaling_config
+
+@functools.lru_cache(maxsize=None)
```

Comment:
Ooooh plugin caching is actually very tricky. We recently implemented an improved decorator which does not require you to implement your own cache clearing mechanism: https://docs.tutor.edly.io/reference/api/hooks/index.html#tutor.hooks.lru_cache

Nice catch, though!